### PR TITLE
stop injecting special discovery behavior into oc/kubectl

### PR DESCRIPTION
**Pull request by deads2k**
Just a POC for now, but this would move `oc` to use nothing but groupified APIs and allow us to eliminate any special factory handling. We'd have fewer special cases.

@openshift/sig-master

### CODE REVIEW

**Comment by deads2k** [comment]
Wow, that's a green test-cmd!

**Comment by deads2k** [comment]
And the e2es are failing on the test for old defaulting. This is great.  Looks like I just need to limit restmapper patch.

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory_client_access.go [hunk 1]
```diff
@@ -111,6 +114,11 @@ func (f *discoveryFactory) DiscoveryClient() (discovery.CachedDiscoveryInterface
 		return nil, err
 	}
 
+	// The more groups you have, the more discovery requests you need to make.
+	// given 25 groups (our groups + a few custom resources) with one-ish version each, discovery needs to make 50 requests
+	// double it just so we don't end up here again for a while.  This config is only used for discovery.
+	cfg.Burst = 100
```
Comment (soltysh) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory_client_access.go]: This definitely deserves upstreaming.

Comment (deads2k) [reply vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/factory_client_access.go]: > This definitely deserves upstreaming.

Already working it

File: test/cmd/authentication.sh [hunk 2]
```diff
@@ -81,8 +81,8 @@ os::cmd::expect_success_and_text "curl -k -XPOST -H 'Content-Type: application/j
 os::cmd::expect_success_and_text "curl -k -XPOST -H 'Content-Type: application/json' -H 'Authorization: Bearer ${accesstoken}' '${API_SCHEME}://${API_HOST}:${API_PORT}/apis/authorization.openshift.io/v1/subjectaccessreviews' -d '{\"namespace\":\"${project}\",\"verb\":\"create\",\"resource\":\"pods\"}'" '"kind": "SubjectAccessReviewResponse"'
 os::cmd::expect_success_and_text "oc policy can-i create pods --token='${accesstoken}' -n '${project}' --ignore-scopes" 'yes'
 os::cmd::expect_success_and_text "oc policy can-i create pods --token='${accesstoken}' -n '${project}'" 'no'
-os::cmd::expect_success_and_text "oc policy can-i create subjectaccessreviews --token='${accesstoken}' -n '${project}'" 'no'
-os::cmd::expect_success_and_text "oc policy can-i create subjectaccessreviews --token='${accesstoken}' -n '${project}' --ignore-scopes" 'yes'
+os::cmd::expect_success_and_text "oc policy can-i create subjectaccessreviews.authorization.openshift.io --token='${accesstoken}' -n '${project}'" 'no'
```
Comment (mfojtik) [inline test/cmd/authentication.sh]: is the full name required now? looks ugly

**Comment by deads2k** [comment]
This is a test update and commit fix up from green.  I think we should do it.

@smarterclayton @liggitt anyone really against?

File: test/cmd/login.sh [hunk 3]
```diff
@@ -27,7 +27,7 @@ cp "${KUBECONFIG}" "${login_kubeconfig}"
 unset KUBECONFIG
 unset KUBERNETES_MASTER
 # test client not configured
-os::cmd::expect_failure_and_text "env -u KUBERNETES_SERVICE_HOST oc get services" 'Missing or incomplete configuration info.  Please login'
+os::cmd::expect_failure_and_text "env -u KUBERNETES_SERVICE_HOST oc get services --loglevel=8" 'Missing or incomplete configuration info.  Please login'
```
Comment (soltysh) [inline test/cmd/login.sh]: Debug leftover?

File: pkg/oc/admin/diagnostics/diagnostics/client/config_contexts.go [hunk 4]
```diff
@@ -235,26 +236,30 @@ func (d ConfigContext) Check() types.DiagnosticResult {
...
-			err = projerr
-		} else { // success!
-			list := []string{}
-			for i, project := range projects.Items {
-				if i > 9 {
-					list = append(list, "...")
-					break
+		projectClient, err := projectclientinternal.NewForConfig(clientConfig)
+		// client create now *fails* if cannot connect to server; so, address connectivity errors below
+		if err == nil {
```
Comment (mfojtik) [inline pkg/oc/admin/diagnostics/diagnostics/client/config_contexts.go]: what happens when err != nil ?

File: pkg/oc/admin/prune/deployments.go [hunk 5]
```diff
@@ -103,11 +103,15 @@ func (o *PruneDeploymentsOptions) Complete(f *clientcmd.Factory, cmd *cobra.Comm
...
 	if err != nil {
 		return err
 	}
-	o.AppsClient = appsClient.Apps()
+	appsClient, err := appsclientinternal.NewForConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+	o.AppsClient = appsClient
 	o.KClient = kClient
```
Comment (mfojtik) [inline pkg/oc/admin/prune/deployments.go]: can you rename this to KubeClient since you are here?

**Comment by mfojtik** [comment]
LGTM as well, nice job!

File: pkg/oc/admin/diagnostics/diagnostics/client/config_contexts.go (hunk 4 above)
Comment (deads2k) [reply pkg/oc/admin/diagnostics/diagnostics/client/config_contexts.go]: > what happens when err != nil ?

Review with whitespace ignored.  This increased nesting.  The error is handled below as it was before.

**Comment by deads2k** [comment]
nits addressed

**Pull request merged**